### ISSUE TITLE
Error deploying contract with hardhat artifact: "invalid type: map, expected valid abi spec file"

### ISSUE THREAD

**Issue by sveitser** [gakonst/ethers-rs#683]
I'm unable to sucessfully use hardhat json output with ethers rs. I did attach a debugger and from what I can tell until the panic occurs when trying to instantiate the contract at the end everything looks good.

**Version**
```
├── ethers v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   ├── ethers-contract v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   ├── ethers-contract-abigen v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   ├── ethers-contract-derive v0.6.0 (proc-macro) (https://github.com/gakonst/ethers-rs#4c677933)
│   │   │   ├── ethers-contract-abigen v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   │   ├── ethers-providers v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   ├── ethers-etherscan v0.2.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   ├── ethers-middleware v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   ├── ethers-contract v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   │   ├── ethers-etherscan v0.2.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   │   ├── ethers-providers v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   │   ├── ethers-signers v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   ├── ethers-providers v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   ├── ethers-signers v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
│   └── ethers-solc v0.1.0 (https://github.com/gakonst/ethers-rs#4c677933)
│       ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
└── ethers-solc v0.1.0 (https://github.com/gakonst/ethers-rs#4c677933) (*)
```
**Platform**
Linux bix 5.10.81 #1-NixOS SMP Sun Nov 21 12:46:37 UTC 2021 x86_64 GNU/Linux

**Description**

I tried this code:

- Deploying to `geth` dev chain.
- `Greeter.sol` is the default hardhat contract.
- Compiled with `hardhat compile`, solc 0.8.8
- `main.rs`
```rust
use anyhow::Result;
use std::{convert::TryFrom, fs, sync::Arc, time::Duration};
use ethers::prelude::*;

abigen!(
    Greeter,
    "./artifacts/contracts/Greeter.sol/Greeter.json",
    event_derives(serde::Deserialize, serde::Serialize);
);

#[tokio::main]
async fn main() -> Result<()> {
    let rpc_url = "http://localhost:8545".to_string();

    let provider = Provider::<Http>::try_from(rpc_url)
        .expect("could not instantiate HTTP Provider")
        .interval(Duration::from_millis(100u64));

    let chain_id = provider.get_chainid().await.unwrap().as_u64();

    // fund deployer account
    let coinbase = provider.get_accounts().await.unwrap()[0];
    let deployer_wallet = LocalWallet::new(&mut rand::thread_rng()).with_chain_id(chain_id); // XXX setting chain_id seems to be required

    let tx = TransactionRequest::new()
        .to(deployer_wallet.address())
        .value(u64::pow(10, 18))
        .from(coinbase);

    provider.send_transaction(tx, None).await?.await?;

    let client = Arc::new(SignerMiddleware::new(provider, deployer_wallet.clone()));

    let path = "./artifacts/contracts/Greeter.sol/Greeter.json";

    let artifact: HardhatArtifact = serde_json::from_reader(fs::File::open(path).unwrap()).unwrap();

    let (abi, bin, _) = artifact.into_parts();

    let factory = ContractFactory::new(abi.unwrap(), bin.unwrap(), client.clone());
    let contract = factory
        .deploy("hello".to_string())?
        .send()
        .await?;

    let _c = Greeter::new(contract.address(), client); // XXX panics!

    Ok(())
}
```

I expected to see this happen: `cargo run` finishes without error

Instead, this happened: panicks with stacktrace

```
thread 'main' panicked at 'invalid abi: Error("invalid type: map, expected valid abi spec file", line: 2, column: 2)', src/main.rs:5:1
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::result::unwrap_failed
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1616:5
   3: core::result::Result<T,E>::expect
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1258:23
   4: hardhat_ethers_rs_solc::greeter_mod::GREETER_ABI::{{closure}}
             at ./src/main.rs:5:1
   5: core::ops::function::FnOnce::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
   7: once_cell::sync::Lazy<T,F>::force::{{closure}}
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/lib.rs:1115:28
   8: once_cell::sync::OnceCell<T>::get_or_init::{{closure}}
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/lib.rs:938:57
   9: once_cell::imp::OnceCell<T>::initialize::{{closure}}
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/imp_std.rs:95:19
  10: once_cell::imp::initialize_inner
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/imp_std.rs:171:31
  11: once_cell::imp::OnceCell<T>::initialize
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/imp_std.rs:93:9
  12: once_cell::sync::OnceCell<T>::get_or_try_init
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/lib.rs:978:13
  13: once_cell::sync::OnceCell<T>::get_or_init
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/lib.rs:938:19
  14: once_cell::sync::Lazy<T,F>::force
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/lib.rs:1114:13
  15: <once_cell::sync::Lazy<T,F> as core::ops::deref::Deref>::deref
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.8.0/src/lib.rs:1124:13
  16: hardhat_ethers_rs_solc::greeter_mod::Greeter<M>::new
             at ./src/main.rs:5:1
  17: hardhat_ethers_rs_solc::main::{{closure}}
             at ./src/main.rs:47:14
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  19: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/park/thread.rs:263:54
  20: tokio::coop::with_budget::{{closure}}
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:106:9
  21: std::thread::local::LocalKey<T>::try_with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:399:16
  22: std::thread::local::LocalKey<T>::with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:375:9
  23: tokio::coop::with_budget
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:99:5
  24: tokio::coop::budget
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:76:5
  25: tokio::park::thread::CachedParkThread::block_on
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/park/thread.rs:263:31
  26: tokio::runtime::enter::Enter::block_on
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/enter.rs:151:13
  27: tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/mod.rs:77:9
  28: tokio::runtime::Runtime::block_on
             at /home/lulu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/mod.rs:463:43
  29: hardhat_ethers_rs_solc::main
             at ./src/main.rs:49:5
  30: core::ops::function::FnOnce::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by lerencao** [comment]
I have come across the same issue.
I think the problem is that `Contract` struct of `ethabi` crate cannot be deserialized from abi json generated by hardhat.
It expects a seq, not a map.

**Comment by gakonst** [comment]
Fixed in #740

**Comment by sveitser** [comment]
This is fixed for the example I posted but when I use hardhat json with `abigen!` the solidity structs are not generated. I can see them missing in the macro expansion of `abigen!` in my real project. Unfortunately can't get the macro expansion to work right now in the project I made to recreate the issue below.

To recreate add a struct to Greeter.sol

```solidity
pragma solidity >=0.8.0;

contract Greeter {

    struct Stuff {
        bool attr;
    }

    function greet(Stuff memory stuff) public view returns (Stuff memory) {
        return stuff;
    }
}
```
and use `main.rs`:
```rust
use anyhow::Result;
use ethers::prelude::*;

abigen!(
    Greeter,
    "./artifacts/contracts/Greeter.sol/Greeter.json",
    event_derives(serde::Deserialize, serde::Serialize);
);

#[tokio::main]
async fn main() -> Result<()> {
    let stuff = Stuff { a: true };
    Ok(())
}
```


```
$ hardhat compile
...
$  cargo run 
   Compiling hardhat-ethers-rs-solc v0.1.0 (/home/lulu/r/sveitser/hardhat-ethers-rs-solc)
error[E0422]: cannot find struct, variant or union type `Stuff` in this scope
  --> src/main.rs:21:17
   |
21 |     let stuff = Stuff { a: true };
   |                 ^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0422`.
error: could not compile `hardhat-ethers-rs-solc` due to previous error
```

**Comment by sveitser** [comment]
The hardhat json

```
{
  "_format": "hh-sol-artifact-1",
  "contractName": "Greeter",
  "sourceName": "contracts/Greeter.sol",
  "abi": [
    {
      "inputs": [
        {
          "components": [
            {
              "internalType": "bool",
              "name": "attr",
              "type": "bool"
            }
          ],
          "internalType": "struct Greeter.Stuff",
          "name": "stuff",
          "type": "tuple"
        }
      ],
      "name": "greet",
      "outputs": [
        {
          "components": [
            {
              "internalType": "bool",
              "name": "attr",
              "type": "bool"
            }
          ],
          "internalType": "struct Greeter.Stuff",
          "name": "",
          "type": "tuple"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    }
  ],
  "bytecode": "0x608060405234801561001057600080fd5b50610230806100206000396000f3fe608060405234801561001057600080fd5b506004361061002b5760003560e01c80635581701b14610030575b600080fd5b61004a600480360381019061004591906100d2565b610060565b6040516100579190610126565b60405180910390f35b610068610070565b819050919050565b60405180602001604052806000151581525090565b600081359050610094816101e3565b92915050565b6000602082840312156100ac57600080fd5b6100b66020610141565b905060006100c684828501610085565b60008301525092915050565b6000602082840312156100e457600080fd5b60006100f28482850161009a565b91505092915050565b61010481610166565b82525050565b60208201600082015161012060008501826100fb565b50505050565b600060208201905061013b600083018461010a565b92915050565b600061014b61015c565b90506101578282610172565b919050565b6000604051905090565b60008115159050919050565b61017b826101d2565b810181811067ffffffffffffffff8211171561019a576101996101a3565b5b80604052505050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6000601f19601f8301169050919050565b6101ec81610166565b81146101f757600080fd5b5056fea2646970667358221220fe505865a0fe8737d5c692ddeebaf3593bb7baa4a0734db6bf9325e6b62bc99964736f6c63430008040033",
  "deployedBytecode": "0x608060405234801561001057600080fd5b506004361061002b5760003560e01c80635581701b14610030575b600080fd5b61004a600480360381019061004591906100d2565b610060565b6040516100579190610126565b60405180910390f35b610068610070565b819050919050565b60405180602001604052806000151581525090565b600081359050610094816101e3565b92915050565b6000602082840312156100ac57600080fd5b6100b66020610141565b905060006100c684828501610085565b60008301525092915050565b6000602082840312156100e457600080fd5b60006100f28482850161009a565b91505092915050565b61010481610166565b82525050565b60208201600082015161012060008501826100fb565b50505050565b600060208201905061013b600083018461010a565b92915050565b600061014b61015c565b90506101578282610172565b919050565b6000604051905090565b60008115159050919050565b61017b826101d2565b810181811067ffffffffffffffff8211171561019a576101996101a3565b5b80604052505050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6000601f19601f8301169050919050565b6101ec81610166565b81146101f757600080fd5b5056fea2646970667358221220fe505865a0fe8737d5c692ddeebaf3593bb7baa4a0734db6bf9325e6b62bc99964736f6c63430008040033",
  "linkReferences": {},
  "deployedLinkReferences": {}
}
```

**Comment by mattsse** [comment]
vote close @gakonst, fixed in #1030
